File: frontend/src/pages/Query5.js
import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';

const Query5 = () => {
    const navigate = useNavigate();
    const [startDate, setStartDate] = useState('')
    const [endDate, setEndDate] = useState('')
    const [pcfViolations, setPCFViolations] = useState([])

    const handleStartDateChange = (event) => {
        setStartDate(event.target.value)
    }

    const handleEndDateChange = (event) => {
        const selectedEndDate = event.target.value;
        if (selectedEndDate >= startDate) {
            setEndDate(selectedEndDate)
        }
    }

    const handlePCFViolationChange = (event) => {
        const selectedPCFViolation = event.target.value
        setPCFViolations(prevViolations => {
            if (prevViolations.includes(selectedPCFViolation)) {
                return prevViolations.filter(violation => violation !== selectedPCFViolation)
            } else {
                return [...prevViolations, selectedPCFViolation]
            }
        })
    }

    const handleSubmit = async (event) => {
        event.preventDefault();

        if (pcfViolations.length === 0) {
            alert('Please select at least one pcf violation.');
            return;
        }

        const data = {
            startDate,
            endDate,
            pcfViolations,
            fromQuery: 'query5'
        }
        setStartDate('')
        setEndDate('')
        setPCFViolations([])
        navigate('/result', {
            state: { data }
        })
       
    }

    return (
        <div className='page'>
            <div className='page-text'>
                <Link to='/'>
                <span className="material-symbols-outlined" title='Back to Welcome Page'>undo</span>
                </Link> 
            <h2>
            Evaluates how PCF violations, such as driving under the influence, have grown more or less common over time in Los Angeles county.
            </h2>
            </div>
            <div className='selection-container'>
                {
                <div className='selection'>
                <h2>PCF Violations: </h2>
                    <label><input type="checkbox" value="1" onChange={handlePCFViolationChange} /> Driving or Bicylcing Under the Influence of Alcohol or Drug</label>
                    <label><input type="checkbox" value="2" onChange={handlePCFViolationChange} /> Impeding Traffic</label>
                    <label><input type="checkbox" value="3" onChange={handlePCFViolationChange} /> Unsafe Speed</label>
                    <label><input type="checkbox" value="4" onChange={handlePCFViolationChange} /> Following Too Closely</label>
                    <label><input type="checkbox" value="5" onChange={handlePCFViolationChange} /> Wrong Side of Road</label>
                    <label><input type="checkbox" value="6" onChange={handlePCFViolationChange} /> Improper Passing</label>
                    <label><input type="checkbox" value="7" onChange={handlePCFViolationChange} /> Unsafe Lane Change</label>
                    <label><input type="checkbox" value="8" onChange={handlePCFViolationChange} /> Improper Turning</label>
                    <label><input type="checkbox" value="9" onChange={handlePCFViolationChange} /> Automobile Right of Way</label>
                    <label><input type="checkbox" value="10" onChange={handlePCFViolationChange} /> Pedestrian Right of Way</label>
                    <label><input type="checkbox" value="11" onChange={handlePCFViolationChange} /> Pedestrian Violation</label>
                    <label><input type="checkbox" value="12" onChange={handlePCFViolationChange} /> Traffic Signals and Signs</label>
                    <label><input type="checkbox" value="13" onChange={handlePCFViolationChange} /> Hazardous Parking</label>
                    <label><input type="checkbox" value="14" onChange={handlePCFViolationChange} /> Lights</label>
                    <label><input type="checkbox" value="15" onChange={handlePCFViolationChange} /> Brakes</label>
                    <label><input type="checkbox" value="16" onChange={handlePCFViolationChange} /> Other Equipment</label>
                    <label><input type="checkbox" value="17" onChange={handlePCFViolationChange} /> Other Hazardous Violation</label>
                    <label><input type="checkbox" value="18" onChange={handlePCFViolationChange} /> Other Than Driver (or Pedestrian)</label>
                    <label><input type="checkbox" value="21" onChange={handlePCFViolationChange} /> Unsafe Starting or Backing</label>
                    <label><input type="checkbox" value="22" onChange={handlePCFViolationChange} /> Other Improper Driving</label>
                    <label><input type="checkbox" value="23" onChange={handlePCFViolationChange} /> Pedestrian or Other Under the Influence of Alcohol or Drug</label>
                    <label><input type="checkbox" value="24" onChange={handlePCFViolationChange} /> Fell Asleep</label>
                    <label><input type="checkbox" value="0" onChange={handlePCFViolationChange} /> Unknown</label>
                </div>
                }
                <div className='time-interval-container selection'>
                    <h2>Time Interval: </h2>
                    <div>
                        <label htmlFor="start-date">Start Date:</label>
                        <input type="date" id="start-date" name="start-date" value={startDate} onChange={handleStartDateChange} min="2013-01-01" max="2022-12-31" />
                    </div>
                    <div>
                        <label htmlFor="end-date">End Date:</label>
                        <input type="date" id="end-date" name="end-date" value={endDate} onChange={handleEndDateChange} min="2013-01-01" max="2022-12-31" />
                    </div>
                </div>
            </div>
                <Link to='/result'>
                    <button className='button' onClick={handleSubmit}>Generate Results</button>
                </Link>
        </div>
    )
}

export default Query5;

File: backend/Query5.py
import cx_Oracle
import pandas as pd
import matplotlib.pyplot as plt

# Database connection details
username = 'greeneryan'
password = 'v51VjDsYBoITgSY0FBrv18sg'
dsn = 'oracle.cise.ufl.edu/orcl'
connection = cx_Oracle.connect(username, password, dsn)
cursor = connection.cursor()

def execute_query(sql, params=None):
    cursor.execute(sql, params or {})
    columns = [col[0] for col in cursor.description]
    data = pd.DataFrame(cursor.fetchall(), columns=columns)
    return data

def get_user_input():
    print("Available PCF Violation Codes and Descriptions:")
    violations = {
        "-": "Not Stated",
        "00": "Unknown",
        "01": "Driving or Bicycling Under the Influence of Alcohol or Drug",
        "02": "Impeding Traffic",
        "03": "Unsafe Speed",
        "04": "Following Too Closely",
        "05": "Wrong Side of Road",
        "06": "Improper Passing",
        "07": "Unsafe Lane Change",
        "08": "Improper Turning",
        "09": "Automobile Right of Way",
        "10": "Pedestrian Right of Way",
        "11": "Pedestrian Violation",
        "12": "Traffic Signals and Signs",
        "13": "Hazardous Parking",
        "14": "Lights",
        "15": "Brakes",
        "16": "Other Equipment",
        "17": "Other Hazardous Violation",
        "18": "Other Than Driver (or Pedestrian)",
        "19": "N/A",
        "20": "N/A",
        "21": "Unsafe Starting or Backing",
        "22": "Other Improper Driving",
        "23": "Pedestrian or Other Under the Influence of Alcohol or Drug",
        "24": "Fell Asleep"
    }
    for code, description in violations.items():
        print(f"{code}: {description}")
    selected_violations = input("Enter violation codes separated by commas (e.g., 01, 03, 05): ").split(', ')
    start_year = input("Enter start year (2013-2022, leave blank for no limit): ")
    end_year = input("Enter end year (2013-2022, leave blank for no limit): ")

    return selected_violations, start_year, end_year

def plot_data(data, violations):
    if data.empty:
        print("No data available to plot.")
        return
    
    plt.figure(figsize=(10, 5))
    for violation in data['VIOLATIONCATEGORY'].unique():
        subset = data[data['VIOLATIONCATEGORY'] == violation]
        plt.plot(subset['YEAR'], subset['NUM_VIOLATIONS'], label=violations[violation])

    plt.title("Trends of Various PCF Violations Over Time")
    plt.xlabel('Year')
    plt.ylabel('Frequency of Violations')
    plt.legend(title="Violation Codes")
    plt.grid(True)
    plt.show()

def main():
    selected_violations, start_year, end_year = get_user_input()
    violations = {
        "-": "Not Stated",
        "00": "Unknown",
        "01": "Driving or Bicycling Under the Influence of Alcohol or Drug",
        "02": "Impeding Traffic",
        "03": "Unsafe Speed",
        "04": "Following Too Closely",
        "05": "Wrong Side of Road",
        "06": "Improper Passing",
        "07": "Unsafe Lane Change",
        "08": "Improper Turning",
        "09": "Automobile Right of Way",
        "10": "Pedestrian Right of Way",
        "11": "Pedestrian Violation",
        "12": "Traffic Signals and Signs",
        "13": "Hazardous Parking",
        "14": "Lights",
        "15": "Brakes",
        "16": "Other Equipment",
        "17": "Other Hazardous Violation",
        "18": "Other Than Driver (or Pedestrian)",
        "19": "N/A",
        "20": "N/A",
        "21": "Unsafe Starting or Backing",
        "22": "Other Improper Driving",
        "23": "Pedestrian or Other Under the Influence of Alcohol or Drug",
        "24": "Fell Asleep"
    }
    sql_query = f"""
    SELECT EXTRACT(YEAR FROM CRASHDATE) AS YEAR, VIOLATIONCATEGORY, COUNT(*) AS NUM_VIOLATIONS
    FROM zhouxiangyu.CRASHES
    WHERE VIOLATIONCATEGORY IN ({','.join(':violation' + str(i) for i in range(len(selected_violations)))})
    """
    if start_year:
        sql_query += " AND EXTRACT(YEAR FROM CRASHDATE) >= :start_year"
    if end_year:
        sql_query += " AND EXTRACT(YEAR FROM CRASHDATE) <= :end_year"
    sql_query += " GROUP BY EXTRACT(YEAR FROM CRASHDATE), VIOLATIONCATEGORY ORDER BY YEAR"

    params = {f'violation{i}': violation for i, violation in enumerate(selected_violations)}
    if start_year:
        params['start_year'] = start_year
    if end_year:
        params['end_year'] = end_year

    data = execute_query(sql_query, params)
    plot_data(data, violations)

    cursor.close()
    connection.close()

if __name__ == '__main__':
    main()
